source/slang/slang.cpp

```diff
@@ -4319,126 +4467,263 @@ String getFileNameFromModuleName(Name* name, bool translateUnderScore)
+    SourceLoc const& requestingLoc,
     DiagnosticSink* sink,
     const LoadedModuleDictionary* loadedModules)
 {
     // Have we already loaded a module matching this name?
     //
-    RefPtr<LoadedModule> loadedModule;
-    if (mapNameToLoadedModules.tryGetValue(name, loadedModule))
+    RefPtr<LoadedModule> previouslyLoadedModule;
+    if (mapNameToLoadedModules.tryGetValue(moduleName, previouslyLoadedModule))
     {
         // If the map shows a null module having been loaded,
         // then that means there was a prior load attempt,
         // but it failed, so we won't bother trying again.
         //
-        if (!loadedModule)
+        if (!previouslyLoadedModule)
             return nullptr;
 
         // If state shows us that the module is already being
         // imported deeper on the call stack, then we've
         // hit a recursive case, and that is an error.
         //
-        if (isBeingImported(loadedModule))
+        if (isBeingImported(previouslyLoadedModule))
         {
             // We seem to be in the middle of loading this module
-            sink->diagnose(loc, Diagnostics::recursiveModuleImport, name);
+            sink->diagnose(requestingLoc, Diagnostics::recursiveModuleImport, moduleName);
             return nullptr;
         }
 
-        return loadedModule;
+        return previouslyLoadedModule;
     }
 
     // If the user is providing an additional list of loaded modules, we find
     // if the module being imported is in that list. This allows a translation
     // unit to use previously checked translation units in the same
     // FrontEndCompileRequest.
-    Module* previouslyLoadedModule = nullptr;
-    if (loadedModules && loadedModules->tryGetValue(name, previouslyLoadedModule))
     {
-        return previouslyLoadedModule;
+        Module* previouslyLoadedLocalModule = nullptr;
+        if (loadedModules && loadedModules->tryGetValue(moduleName, previouslyLoadedLocalModule))
+        {
+            return previouslyLoadedLocalModule;
+        }
     }
 
-    if (name == getSessionImpl()->glslModuleName)
+    // If the name being requested matches the name of a built-in module,
+    // then we will special-case the process by loading that builtin
+    // module directly.
+    //
+    // TODO: right now this logic is only considering the built-in `glsl`
+    // module, but it should probably be generalized so that we can more
+    // easily support having multiple built-in modules rather than just
+    // putting everything into `core`.
+    //
+    if (moduleName == getSessionImpl()->glslModuleName)
     {
         // This is a builtin glsl module, just load it from embedded definition.
         auto glslModule = getSessionImpl()->getBuiltinModule(slang::BuiltinModuleName::GLSL);
         if (!glslModule)
         {
-            sink->diagnose(loc, Diagnostics::glslModuleNotAvailable, name);
+            // Note: the way this logic is currently written, if the built-in
```

Comment:
This is an interesting question. Allowing user modules to shadow builtin module seems useful, but it could also lead to confusions. I am not sure what is the best answer here.